modules/sw360/sw360-client/src/main/java/org/eclipse/sw360/antenna/sw360/client/rest/SW360Client.java

```diff
@@ -0,0 +1,263 @@
+     * @return a future with the result generated by the {@code ResponseProcessor}
+     */
+    protected <T> CompletableFuture<T> executeRequest(RequestProducer producer, ResponseProcessor<T> processor,
+                                                      String tag) {
+        return executeWithExpiredTokenCheck(producer, processor, tag, true);
+    }
+
+    /**
+     * Executes a request to SW360 with authentication and retry logic that
+     * expects a JSON response. This method performs the same checks as
+     * {@code executeRequest()}, but it creates the {@code ResponseProcessor}
+     * automatically that can convert the JSON response to the desired target
+     * type.
+     *
+     * @param producer    the {@code RequestProducer}
+     * @param resultClass the class to which the JSON payload is to be
+     *                    converted
+     * @param tag         a tag to identify the request
+     * @param <T>         the type of the result
+     * @return a future with the result of the request
+     */
+    protected <T> CompletableFuture<T> executeJsonRequest(RequestProducer producer, Class<T> resultClass, String tag) {
+        return executeRequest(producer,
+                HttpUtils.jsonResult(getClientConfig().getObjectMapper(), resultClass), tag);
+    }
+
+    /**
+     * Generates a URL pointing to a specific resource of the SW360 server.
+     * This method concatenates the given path segments and appends them to the
+     * base REST URL of the server. It can be used by derived classes to
+     * generate the URLs to be requested.
+     *
+     * @param paths the path segments to add to the base URL
+     * @return a string with the resulting URL
+     */
+    protected String resourceUrl(String... paths) {
+        return getClientConfig().getRestURL() + URL_PATH_SEPARATOR + String.join(URL_PATH_SEPARATOR, paths);
+    }
+
+    /**
+     * Implements the actual request execution logic including a retry
+     * mechanism if the current access token may have expired. This method
+     * generates a future that obtains the current access token and executes
+     * the request defined by the parameters with it. If the {@code canRetry}
+     * parameter is <strong>true</strong> (which it is only for the first time
+     * the request is executed), the future is enhanced with special retry
+     * logic. This logic checks whether a failure has occurred indicating an
+     * expired access token. If this is the case, another request execution is
+     * chained to the original future.
+     *
+     * @param producer  the {@code RequestProducer}
+     * @param processor the {@code ResponseProcessor}
+     * @param tag       a tag to identify the request
+     * @param canRetry  a flag whether a retry is possible
+     * @param <T>       the type of the result
+     * @return a future with the result of the request
+     */
+    private <T> CompletableFuture<T> executeWithExpiredTokenCheck(RequestProducer producer,
```

Comment:
I am having issues with this method. First the name indicates a special execution of requests, but it is the standard approach to run requests. The original intention of running the request is hard to understand because of this embedding into the token check. I have no idea how to improve, but it took a while until I understood. Also the name of the withConditionalFallback method is from the name not helping to be honest. Perhaps some name like createFutureWrapperForConditionalFallback or something like that.

Reply:
The methods have been renamed, so that their intension is hopefully clearer.